rustls/src/server/tls13.rs

```diff
@@ -880,23 +880,38 @@ impl ExpectFinished {
         })
     }
 
-    fn emit_stateless_ticket(
+    fn emit_ticket(
         handshake: &mut HandshakeDetails,
         sess: &mut ServerSessionImpl,
         key_schedule: &KeyScheduleTraffic,
     ) {
         let nonce = rand::random_vec(32);
         let plain = get_server_session_value(handshake, key_schedule, sess, &nonce).get_encoding();
-        let ticket = match sess.config.ticketer.encrypt(&plain) {
-            Some(t) => t,
-            None => return,
+
+        let stateless = sess.config.ticketer.enabled();
+        let (ticket, lifetime) = if stateless {
+            let ticket = match sess.config.ticketer.encrypt(&plain) {
+                Some(t) => t,
+                None => return,
+            };
+            (ticket, sess.config.ticketer.get_lifetime())
+        } else {
+            let id = rand::random_vec(32);
+            let stored = sess
+                .config
+                .session_storage
+                .put(id.clone(), plain);
+            if !stored {
+                trace!("resumption not available; not issuing ticket");
+                return;
+            }
+            let stateful_lifetime = 24 * 60 * 60; // this is a bit of a punt
+            (id, stateful_lifetime)
         };
-        let ticket_lifetime = sess.config.ticketer.get_lifetime();
 
-        let age_add = rand::random_u32(); // nb, we don't do 0-RTT data, so whatever
```

Comment:
I don't understand the "nb, we don't do 0-RTT data, so whatever" comment that was removed.

Reply:
This comment is out of date: it was written before rustls-in-quic-mode did 0-RTT data.